omc3/knob_extractor.py

```diff
@@ -0,0 +1,459 @@
+    extracts the chromaticity and dispersion knobs at 14h on May 4th 2022
+
+python knob_extractor.py --knobs disp chroma --time now _2h 
+    extracts the chromaticity and dispersion knobs as of 2 hours ago
+
+python knob_extractor.py --state
+    prints the current StateTracker/State metadata
+
+python knob_extractor.py disp sep xing chroma ip_offset mo --time now
+    extracts the current settings for all the knobs
+"""
+
+
+def get_params():
+    return EntryPointParameters(
+        knobs=dict(
+            type=str,
+            nargs='*',
+            help=(
+                "A list of knob names or categories to extract. "
+                f"Available categories are: {', '.join(KNOB_CATEGORIES.keys())}."
+            ),
+            default=list(KNOB_CATEGORIES.keys()),
+        ),
+        time=dict(
+            type=str,
+            help=(
+                "At what time to extract the knobs. "
+                "Accepts ISO-format (YYYY-MM-DDThh:mm:ss), timestamp or 'now'. "
+                "The default timezone for the ISO-format is local time, "
+                "but you can force e.g. UTC by adding +00:00."
+            ),
+            default="now",
+        ),
+        timedelta=dict(
+            type=str,
+            help=(
+                "Add this timedelta to the given time. "
+                "The format of timedelta is '((\\d+)(\\w))+' "
+                "with the second token being one of "
+                "s(seconds), m(minutes), h(hours), d(days), w(weeks), M(months) "
+                "e.g 7m = 7 minutes, 1d = 1day, 7m30s = 7 min 30 secs. "
+                "A prefix '_' specifies a negative timedelta. "
+                "This allows for easily getting the setting "
+                "e.g. 2h ago: '_2h' while setting the `time` argument to 'now' (default)."
+            ),
+        ),
+        state=dict(
+            action='store_true',
+            help=(
+                "Prints the state of the StateTracker. "
+                "Does not extract anything else."
+            ),
+        ),
+        output=dict(
+            type=PathOrStr,
+            default='knobs.madx',
+            help=(
+                "Specify user-defined output path. "
+                "This should probably be `model_dir/knobs.madx`"
+            ),
+        ),
+        knob_definitions=dict(
+            type=PathOrStrOrDataFrame,
+            help=(
+                "User defined path to the knob-definitions, "
+                "or (via python) a dataframe containing the knob definitions "
+                "with the columns 'madx', 'lsa' and 'scaling'."
+            ),
+        ),
+    )
+
+
+@dataclass
+class KnobEntry:
+    madx: str
+    lsa: str
+    scaling: float  # is usually +-1, i.e. takes care of sign-conventions
+    value: float = None
+
+    def get_madx(self):
+        if self.value is None:
+            return f"! {self.madx} : No Value extracted"
+        return f"{self.madx} := {self.value * self.scaling};"
+
```

Comment:
I like `structs`. but at this point this is just bloat

Reply:
What's the issue with organised data structures?